CONTRIBUTING.md

zbanks opened the pull request:
Moved from #142 after incorporating all of the great feedback. 

The PR template could either go in the root directory, `docs/` or `.github/`. I put it in the `.github/` because we don't have a `docs/` folder yet.

Kenix3 commented:
https://github.com/zeldaret/mm/issues/144

Kenix3 commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,13 @@
+
+- `./format.sh` was run to apply standard formatting.
+- `make` successfully builds a matching ROM.
+- No new compiler warnings during the build process
```
```suggestion
- No new compiler warnings were introduced during the build process
```

Kenix3 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,147 @@
+After all the comments are addressed and at least one contributor has approved the review, the project lead can then merge the code.
+The project lead is also responsible for ensuring that all of these procedures are followed.
+
+Throughout the process, you (the PR author) should update the [Trello board](https://trello.com/b/ruxw9n6m/majoras-mask-decompilation) with the status of your PR.
```
Updating trello includes both putting it in the correct column, and updating it with appropriate labels.

zbanks commented:
If anyone wants to see what the template looks like, I put this branch on my fork's master, so you can [preview it here](https://github.com/zbanks/mm/compare/master...zbanks:fix-warnings).

AngheloAlf commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,147 @@
+    - Global variables start with `g` (`gSaveContext`)
+    - Static global variables start with `s` (`sSphereInit`)
+- Macros and enum constants are `SCREAMING_SNAKE_CASE` (`DM_RAVINE_STATE_ACTIVE`)
+- Auxiliary structs, enums, and constants are defined in the header file
```
There are some structs, enums or constants that are only related to the internal use of the an actor, so moving them to the header doesn't make much sense to me.

I would suggest only defining those stuff in a header if it does make sense using them _outside_ of that specific actor. For example, an enum which has the values used in `actor.params`.

zbanks replied:
I'd tend to agree, that would have been my preference, but I could have sworn someone had told me to make the state enum public in my #120 PR for Dm_Ravine. (I can't find a record of this though...).

But I'm happy to reword this either way. Do you (or does anyone else) have a succinct way of describing this? I'm having trouble.
I think normally I'd call them "public structs" or "structs that are part of the API" but I think those terms aren't very clear here because we don't use them anywhere else in the project.

AngheloAlf commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,147 @@
+- Macros and enum constants are `SCREAMING_SNAKE_CASE` (`DM_RAVINE_STATE_ACTIVE`)
+- Auxiliary structs, enums, and constants are defined in the header file
+- Trailing commas in array and struct definitions (see `EnJcMatoDamageTable`)
+- Constants converted to whichever looks best in context: hexadecimal, decimal, float, or boolean
```
May be worth noting we don't allow the use of the type `bool`, but we can use `true` and `false` if it makes sense.

zbanks replied:
I'll just drop the mention of boolean entirely; `true`/`false` are pretty rare so far.

AngheloAlf commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,147 @@
+    - Rotation angles should always be in hexadecimal
+- Define macros for bitwise access to `actor->params`
+
+### `NON_MATCHING`
```
We should at least mention the existence of `NON_EQUIVALENT` too.

zbanks replied:
What's the difference? 

I think explaining `NON_MATCHING` is probably good enough. I think the key point I wanted to convey here is that "if you get a function really close but you can't finish it and you're going to give up, consider `NON_MATCHING` instead of throwing it out."

AngheloAlf commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,147 @@
+
+- `./format.sh` was run to apply standard formatting.
+- `make` successfully builds a matching ROM.
+- No new compiler warnings during the build process
```
```suggestion
- No new compiler warnings during the build process.
```

Zelllll commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,147 @@
+
+These files demonstrate the following:
+
+- Word order in names are from least-to-most specific (`DM_RAVINE_STATE_ACTIVE`, not `DM_RAVINE_ACTIVE_STATE`)
```
Please, please do not suggest people to use the og names for enums, variables, etc

Kenix3 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,147 @@
+
+- `./format.sh` was run to apply standard formatting.
+- `make` successfully builds a matching ROM.
+- No new compiler warnings were introduced during the build process.
```
How would someone verify this?
tools/warnings_count/check_new_warnings.sh

Kenix3 commented on README.md:
````diff
@@ -23,7 +23,7 @@ in an experimental and research phase and cannot currently be used traditionally
 ```
 
-This repo does not include all assets necessary for compiling the ROM. A prior copy of the game is required to extract the required assets.
+**This repo does not include any assets or code necessary for compiling the ROM. A prior copy of the game is required to extract the required assets.**
````
```suggestion
**This repo does not include any assets necessary for compiling the ROM. A prior copy of the game is required to extract the required assets.**
```
This is the wording we decided on in Discord.

EllipticEllipsis commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,147 @@
+This document is meant to be a set of tips and guidelines for contributing to the project.
+For general information about the project, see [our readme](https://github.com/zeldaret/mm/blob/master/README.md).
+
+Most discussions happen on our [Discord Server](https://discord.zelda64.dev), where you are welcome to ask if you need help getting started, or if you have any questions regarding this project and other decompilation projects.
```
Isn't this a non-restrictive relative clause, and so conventionally is written with a comma?

EllipticEllipsis commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,147 @@
+
+Basic knowledge of C, particularly arrays and pointers, is extremely useful. Knowledge of MIPS is not required initially, you can pick it up as you work through some decompilation.
+
+You should be familiar with using git and GitHub. There are a number of tutorials available online, [such as this one](https://github.com/firstcontributions/first-contributions), which can help you get started.
```
And here, "such as this one" is distinct from the subsequent relative clause: the linked one is not the only tutorial that can help you get started.

EllipticEllipsis commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,147 @@
+Style Guide & Conventions
+-------------------------
+
+Most of the C formatting style is enforced by the `format.sh` script, which is based on `clang-format`.
```
Also a non-restrictive relative clause.

zbanks replied:
I just removed the commas, I think it's reasonable either way.

zbanks commented:
I think I got all of the fixes you suggested @Kenix3. (I removed the commas, I think it's readable either way)

I also changed the PR template a bit to make it a bit "denser", removing some of the whitespace & combined the two `<!-- ... -->` comments. I've been keeping it in sync with the "checklist" section in the main contributing.md file.

zbanks commented on README.md:
```diff
@@ -103,9 +103,21 @@ This means that something is wrong with the ROM's contents. Either the baserom f
 
 Running `make init` will also make the `./expected` directory and copy all of the files there, which will be useful when running the diff script. The diff script is useful in decompiling functions and can be ran with this command: `./tools/asm-differ/diff.py -wmo3 <insert_function_here>`
 
```
I changed the format slightly to be consistent with the next note (`**Note**:`), and added an example `make -j4` (I think it can be confusing to convey that the `N` is not literal)

Thar0 commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,155 @@
+
+```c
+#ifdef NON_MATCHING
+void CollisionCheck_SpawnWaterDroplets(GlobalContext* ctxt, Vec3f* v) {
````
```suggestion
void CollisionCheck_SpawnWaterDroplets(GlobalContext* globalCtx, Vec3f* v) {
```
Nitpick but we want to phase out all the variations on the global context variable as soon as possible, so might as well have it correct in the docs from the start, code will follow soon enough.

Thar0 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,155 @@
+- Using (or adding) constants, enums, and macros when possible
+- Explaining sections of code that are not straightforward
+
+Overlays are not required to be documented at this time, but `code/` and `boot/` are.
```
I don't really agree that anything should be "required" to be documented. Just having decompiled code is a huge step forward and requiring people who might not have the time to sit down and fully understand the system they are decompiling at the time seems inappropriate. Not to mention, certain core systems that operate in close connection with other systems or make use of subsystems may be unreasonable to document properly at the time and will need to be revisited after all other relevant information is available to ensure the documentation is accurate.

Kenix3 replied:
Good point that there are exceptions when one system calls on another system a lot. There should probably be a carve out exception for those situations. That being said, in most cases that documentation these days can be gleaned from looking at OOT code.

As far as someone not having time, to be honest, my solution to them would be to work on an actor where documentation is not required.

zbanks replied:
Yeah this makes sense. I think avoiding saying "X is required" is a net positive, it sounds a bit rigid.

```suggestion
Overlays are not required to be documented at this time, but `code/` and `boot/` should be documented as much as possible.
```

@Kenix3 what do you think? Do you have any suggestions here?

Kenix3 replied:
I'm happy with that rewording.

Thar0 commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,9 @@
+Before opening this PR, ensure the following:
+- `./format.sh` was run to apply standard formatting.
+- `make` successfully builds a matching ROM.
+- No new compiler warnings were introduced during the build process.
+    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
```
Is there anywhere that mentions to consult in the discord server if it seems like the warnings may be necessary? I suppose people will naturally do this if they struggle with removing certain warnings, but it might be better to just have it written down to let them know that's encouraged. I think it would be better than someone potentially agonising over trying to remove a particularly tough warning.

zbanks replied:
I added it to the contributing.md as a generic note, but not here? Do you think that's fine?
I think in practice people could run into issues at any step, not just the warnings. (e.g. if you don't have matching code)

Thar0 replied:
I agree it's not just limited to that one point so the generic statement is a good idea. I suppose what I was looking for in particular was a note or prompt in the PR template that if there is any particular issues preventing a PR they can be discussed, but if you feel as though that's already made clear enough then it's likely fine.

Thar0 reviewed (commented):
Overall looks great, just a few things.
